### PR TITLE
Use the URL structure to compare the current mailbox with C_Spoolfile

### LINKED ISSUE #2379: $move is broken on imap with imap_user
## Expected Behaviour

With `$move=yes` read mail in my `$spoolfile=+INBOX` should be moved to `$mbox=+READ`.

## Actual Behaviour

In neomutt 20200619 the mails stay in inbox.

## Steps to Reproduce

1. `neomutt -F /dev/null -f imaps://server -e 'set imap_user=username' -e 'set imap_check_subscribed' -e 'set folder="imaps://server"' -e 'set spoolfile=+INBOX' -e 'set mbox=+READ' -e 'set move=yes'`
2. read an email
3. `c` `!` `<enter>`, and the mails aren't moved

## How often does this happen?

Always. But here's a slightly modified command line that does _not_ have the bug:

`neomutt -F /dev/null -f imaps://username@server -e 'set imap_user=username' -e 'set imap_check_subscribed' -e 'set folder="imaps://username@server"' -e 'set spoolfile=+INBOX' -e 'set mbox=+READ' -e 'set move=yes'`

Note the addition of `username@` to folder.

## When did it start to happen?

In neomutt 20200619. The last version I've seen this working in was 20180716, but I haven't tested the versions in between because other bugs forced me to stay with that old version for quite a while.

## NeoMutt Version

```
NeoMutt 20200619
Copyright (C) 1996-2016 Michael R. Elkins and others.
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Linux 5.6.0-2-amd64 (x86_64)
ncurses: ncurses 6.2.20200212 (compiled with 6.2.20200212)
libidn: 1.33 (compiled with 1.33)
GPGME: 1.13.1-unknown
GnuTLS: 3.6.14
libnotmuch: 5.2.0
storage: tokyocabinet

Configure options: --build=x86_64-linux-gnu --prefix=/usr {--includedir=${prefix}/include} {--mandir=${prefix}/share/man} {--infodir=${prefix}/share/info} --sysconfdir=/etc --localstatedir=/var --disable-option-checking --disable-silent-rules {--libdir=${prefix}/lib/x86_64-linux-gnu} {--libexecdir=${prefix}/lib/x86_64-linux-gnu} --disable-maintainer-mode --disable-dependency-tracking --mandir=/usr/share/man --libexecdir=/usr/libexec --with-mailpath=/var/mail --gpgme --lua --notmuch --with-ui --gnutls --gss --idn --mixmaster --sasl --tokyocabinet

Compilation CFLAGS: -g -O2 -fdebug-prefix-map=/build/neomutt-MrazQb/neomutt-20200619+dfsg.1=. -fstack-protector-strong -Wformat -Werror=format-security -std=c99 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/usr/include -I/usr/include/lua5.3 -DNCURSES_WIDECHAR -isystem /usr/include/mit-krb5

Default options:
  +attach_headers_color +compose_to_sender +compress +cond_date +debug 
  +encrypt_to_self +forgotten_attachments +forwref +ifdef +imap +index_color 
  +initials +limit_current_thread +multiple_fcc +nested_if +new_mail +nntp +pop 
  +progress +quasi_delete +regcomp +reply_with_xorig +sensible_browser +sidebar 
  +skip_quoted +smtp +status_color +timeout +tls_sni +trash 

Compile options:
  -autocrypt +bkgdset +color +curs_set +fcntl -flock -fmemopen +futimens 
  +getaddrinfo +gnutls +gpgme +gss +hcache -homespool +idn +inotify 
  -locales_hack +lua +meta +mixmaster +nls +notmuch -openssl +pgp +regex +sasl 
  +smime -sqlite +start_color +sun_attachment +typeahead 
MAILPATH="/var/mail"
MIXMASTER="mixmaster"
PKGDATADIR="/usr/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/etc"

To learn more about NeoMutt, visit: https://neomutt.org
If you find a bug in NeoMutt, please raise an issue at:
    https://github.com/neomutt/neomutt/issues
or send an email to: <neomutt-devel@neomutt.org>
```

## Extra Info

I'm using the latest Debian package of neomutt.

### PR DESCRIPTION
**Pull request by gahr**
Fixes #2379

I'm totally not convinced this is the best way to approach it :/

See https://github.com/neomutt/neomutt/issues/2379#issuecomment-647482328 for the reasons.

### CODE REVIEW

File: mx.c [hunk 1]
```diff
@@ -155,7 +155,28 @@ const struct MxOps *mx_get_ops(enum MailboxType type)
...
+  if (ua->scheme != ub->scheme)
+    goto done;
+  if (mutt_str_strcasecmp(ua->host, ub->host) != 0)
+    goto done;
+  if (mutt_str_strcasecmp(ua->path, ub->path) != 0)
+    goto done;
+  if (ua->user && ub->user && mutt_str_strcmp(ua->user, ub->user) != 0)
+    goto done;
+
+  is_spool = true;
```
Comment (daedric) [inline mx.c:174]: couldn't you simplify by having something such as:
`is_spool = ua->scheme == ub->scheme && mutt_str_strcasecmp(ua->host, ub->host) == 0 && mutt_str_strcasecmp(ua->path, ub->path) == 0 && mutt_str_strcmp(ua->user, ub->user) == 0;` ?
It woudl avoid the goto and mutt_str_strcmp handles null pointer

Comment (gahr) [reply mx.c:174]: I could put scheme, host, and path together, but I only want to compare usernames if they're both non-null.

Comment (daedric) [reply mx.c:174]: fair enough

Comment (gahr) [reply mx.c:174]: Like this: 
```c
  const bool is_spool =                                                             
    (ua->scheme == ub->scheme) &&                                                   
    (mutt_str_strcasecmp(ua->host, ub->host) == 0) &&                               
    (mutt_str_strcasecmp(ua->path, ub->path) == 0) &&                               
    (!ua->user || !ub->user || (mutt_str_strcmp(ua->user, ub->user) == 0));
```

Comment (daedric) [reply mx.c]: I personally find it clearer :)

File: mx.c [hunk 2]
```diff
@@ -155,7 +155,20 @@ const struct MxOps *mx_get_ops(enum MailboxType type)
  */
 static bool mutt_is_spool(const char *str)
 {
-  return mutt_str_strcmp(C_Spoolfile, str) == 0;
+  struct Url *ua = url_parse(str);
+  struct Url *ub = url_parse(C_Spoolfile);
+
+  if (!ua || !ub)
+    return false;
```
Comment (daedric) [inline mx.c:162]: you forgot to cleanup in this case

Comment (daedric) [reply mx.c:162]: (ua or ub could have been created)

**Pull request merged**